multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,287 @@
+		podLabelCache:    map[string]string{},
+	}
+}
+
+// +kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Pod/Namespace for label identity", "pod/ns", req.NamespacedName)
+	var commonArea commonarea.RemoteCommonArea
+	var err error
+	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			return ctrl.Result{}, r.onPodDelete(ctx, req.NamespacedName.String())
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	nsLabels, podLabels := ns.Labels, pod.Labels
+	return ctrl.Result{}, r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), getNormalizedLabel(nsLabels, podLabels, ns.Name))
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Pod{}).
+		WithEventFilter(predicate.LabelChangedPredicate{}).
+		Watches(&source.Kind{Type: &v1.Namespace{}}, handler.EnqueueRequestsFromMapFunc(r.namespaceMapFunc)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: common.DefaultWorkerCount,
+		}).
+		Complete(r)
+}
+
+// namespaceMapFunc handles Namespace update events (Namespace label change) by enqueuing
+// all Pods in the Namespace into the reconciler processing queue.
+func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile.Request {
+	podList := &v1.PodList{}
+	r.Client.List(context.TODO(), podList, client.InNamespace(ns.GetName()))
+	requests := make([]reconcile.Request, len(podList.Items))
+	for i, pod := range podList.Items {
+		requests[i] = reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      pod.GetName(),
+				Namespace: pod.GetNamespace(),
+			},
+		}
+	}
+	return requests
+}
+
+// onPodDelete removes the Pod and label identity mapping from the cache, and
+// updates label identity ResourceExport if necessary (the Pod deletion event
+// causes a label identity no longer present in the cluster).
+func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string) error {
+	r.labelMutex.Lock()
```

Comment:
Do we still need to lock the whole func, or we can go your earlier approach of locking when accessing/updating the label cache?

Reply:
I do think the issue I presented earlier is still valid, and in that scenario it is not an unnecessary creation/deletion call that we can ignore, but a missed label identity add call which will definitely cause issues. In order to fix the issue and only lock label cache at the same time, there needs to be extra logic for checking in-flight label add/delete

Reply:
You are right. Seems the issue is only when add/delete at the same time? I cannot think of any simple solution without locking the funcs. Got ideas like: checking the state after deleting Export and recreating if a new Pod using the ID; or re-queueing the add event if delete is in process; or queueing Export creation/delete for the create/delete calls of a given ID to be handled by a single thread. But all sound complex.

However, if we lock the whole func, does it mean no difference to use a single worker to handle all events, which we assume can have scale issues?

Reply:
I just added a new commit trying to resolve this issue. Essentially it re-queues the add event if either a ResourceExport for this label is being created/deleted. I also added UT coverage for these cases. Hopefully it is not too complex. With that, we label cache updates will not be blocked by API resource update events